models/auth/token_scope.go

```diff
@@ -0,0 +1,230 @@
+	bitmap, err := s.Parse()
+	if err != nil {
+		return "", err
+	}
+
+	return bitmap.ToScope(), nil
+}
+
+// HasScope returns true if the string has the given scope
+func (s AccessTokenScope) HasScope(scope string) (bool, error) {
+	index := sliceIndex(AllAccessTokenScopes, scope)
+	if index == -1 {
+		return false, fmt.Errorf("invalid access token scope: %s", scope)
+	}
+
+	bitmap, err := s.Parse()
+	if err != nil {
+		return false, err
+	}
+
+	return bitmap&(1<<uint(index)) != 0, nil
+}
+
+// ToScope returns a normalized scope string without any duplicates.
+func (bitmap AccessTokenScopeBitmap) ToScope() AccessTokenScope {
+	var scopes []string
+
+	groupedScope := make(map[string]struct{})
+	for i, v := range AllAccessTokenScopes {
+		if bitmap&(1<<uint(i)) != 0 {
+			switch v {
+			// Parse scopes that contains multiple sub-scopes
+			case AccessTokenScopeRepo, AccessTokenScopeAdminOrg, AccessTokenScopeAdminPublicKey,
+				AccessTokenScopeUser, AccessTokenScopePackage, AccessTokenScopeAdminGPGKey:
+				groupedScope[v] = struct{}{}
+			case AccessTokenScopeAdminRepoHook:
+				groupedScope[v] = struct{}{}
+				if _, ok := groupedScope[AccessTokenScopeRepo]; ok {
+					continue
+				}
+
+			// If parent scope is set, all sub-scopes shouldn't be added
+			case AccessTokenScopeRepoStatus, AccessTokenScopePublicRepo:
+				if _, ok := groupedScope[AccessTokenScopeRepo]; ok {
+					continue
+				}
+			case AccessTokenScopeWriteOrg, AccessTokenScopeReadOrg:
+				if _, ok := groupedScope[AccessTokenScopeAdminOrg]; ok {
+					continue
+				}
+			case AccessTokenScopeWritePublicKey, AccessTokenScopeReadPublicKey:
+				if _, ok := groupedScope[AccessTokenScopeAdminPublicKey]; ok {
+					continue
+				}
+			case AccessTokenScopeWriteRepoHook, AccessTokenScopeReadRepoHook:
+				if _, ok := groupedScope[AccessTokenScopeAdminRepoHook]; ok {
+					continue
+				}
+				if _, ok := groupedScope[AccessTokenScopeRepo]; ok {
+					continue
+				}
+			case AccessTokenScopeReadUser, AccessTokenScopeUserEmail, AccessTokenScopeUserFollow:
+				if _, ok := groupedScope[AccessTokenScopeUser]; ok {
+					continue
+				}
+			case AccessTokenScopeWritePackage, AccessTokenScopeReadPackage, AccessTokenScopeDeletePackage:
+				if _, ok := groupedScope[AccessTokenScopePackage]; ok {
+					continue
+				}
+			case AccessTokenScopeWriteGPGKey, AccessTokenScopeReadGPGKey:
+				if _, ok := groupedScope[AccessTokenScopeAdminGPGKey]; ok {
+					continue
+				}
+			}
+			scopes = append(scopes, v)
+		}
+	}
+
+	scope := AccessTokenScope(strings.Join(scopes, ","))
+	scope = AccessTokenScope(strings.ReplaceAll(
+		string(scope),
+		"repo,admin:org,admin:public_key,admin:org_hook,notification,user,delete_repo,package,admin:gpg_key",
+		"all",
+	))
+	return scope
+}
+
+// sliceIndex returns the index of the first instance of str in slice, or -1 if str is not present in slice.
+func sliceIndex(slice []string, element string) int {
```

Comment:
Looks like we already have this function `IsStringInSlice` in `modules/util`

Reply:
I see... But here we need the index of the string, not just if it exists or not. I could add the function to `modules/util` though. What do you think?

Reply:
> I see... But here we need the index of the string, not just if it exists or not. I could add the function to `modules/util` though. What do you think?

Yes, please move it to `modules/util` when possible.